### PR TITLE
API-1486: ES query done

### PR DESCRIPTION
**Pull request by tomglvng**
We add a query to search on Events API debug logs.

We want the logs `warning` and `error` of the last 72h.
We want the last 100 `notice` and `info`.
We want to have those 2 in the same list, ordered by date DESC.

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/Domain/Webhook/Persistence/Query/SearchEventSubscriptionDebugLogsQueryInterface.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+     *    timestamp: int,
+     *    level: string,
+     *    message: string,
+     *    connection_code: ?string,
+     *    context: array
+     *  }>,
+     *  total: int,
+     *  search_after: string
+     *}>
+     */
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Domain/Webhook/Persistence/Query/SearchEventSubscriptionDebugLogsQueryInterface.php:24]: TODO: add comment what is it ?

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    public function __construct(
+        Client $elasticsearchClient,
+        Clock $clock,
+        PrimaryKeyEncrypter $encrypter
+    ) {
+        $this->elasticsearchClient = $elasticsearchClient;
+        $this->clock = $clock;
+        $this->encrypter = $encrypter;
+    }
+
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:34]: TODO: add pseudo-sql for explanation

Comment (tomglvng) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:34]: SELECT id FROM logs
WHERE level IN ['info', 'notice']
LIMIT 100;

SELECT * FROM logs
WHERE (level IN ['warning', 'error']) OR (level IN ['notice', 'info'] AND id IN ids)

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 3]
```diff
@@ -10,6 +10,50 @@
...
+ * The functionnal need of this query is not easy to do in ES.
+ * The query is working and well tested but not easy to read at first glance.
+ *
+ * We want the logs `warning` and `error` of the last 72h.
+ * We want the last 100 `notice` and `info`.
+ * We want to have those 2 in the same list, ordered by date DESC.
+ *
+ * We know it's easier to write its counterpart in pseudo-SQL, so here it is:
+ *
+ * SELECT * FROM logs
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:23]: To remove

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            'sort' => [
+                'timestamp' => 'DESC',
+                '_id' => 'ASC'
+            ],
+            'track_total_hits' => true,
+            'query' => [
+                'bool' => [
+                    'should' => [
+                        ['bool' => ['must' => [
+                            ['terms' => ['level' => ['info', 'notice']]],
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:107]: useless

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+                            ['terms' => ['_id' => $lastNoticeAndInfoIdentifiers['identifiers']]],
+                            ['bool' => ['should' => [
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:109]: useless

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    /**
+     * @return array{
+     *   first_id: ?string,
+     *   first_sort: ?array<int, string>,
+     *   identifiers: array<string>
+     * }
+     */
+    private function findSameLastNoticeAndInfoIdentifiers(string $connectionCode, string $firstId, array $firstSort): array
+    {
+        $result = $this->elasticsearchClient->search(
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:227]: TODO: function for the ES query

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+            }, $result['hits']['hits']),
+        ];
+    }
+
+    /**
+     * @return array{
+     *   first_id: ?string,
+     *   first_sort: ?array<int, string>,
+     *   identifiers: array<string>
+     * }
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:223]: Add explanation

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+    {
+        $result = $this->elasticsearchClient->search(
+            [
+                '_source' => ['id'],
+                'sort' => [
+                    'timestamp' => 'DESC',
+                    '_id' => 'ASC'
+                ],
+                'search_after' => $firstSort,
+                'size' => self::MAX_NUMBER_OF_NOTICE_AND_INFO_LOGS - 1,
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:235]: add explanation

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    public function execute(string $connectionCode, ?string $encryptedSearchAfter = null): array
+    {
+        $parameters = $this->buildParameters($encryptedSearchAfter);
+
+        $nowTimestamp = $this->clock->now()->getTimestamp();
+
+        if (null !== $parameters['first_notice_and_info_id'] && null !== $parameters['first_notice_and_info_search_after']) {
+            $lastNoticeAndInfoIdentifiers = $this->findSameLastNoticeAndInfoIdentifiers(
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:87]: TODO: 'identifier' to 'id'

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQueryIntegration.php [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+                    'message' => 'Foo bar',
+                    'connection_code' => null,
+                    'context' => [],
+                ];
+            },
+            30
+        );
+
+        $firstPage = $this->query->execute('a_connection_code');
+
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQueryIntegration.php:244]: add: assertCount 25

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQueryIntegration.php [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+                    'connection_code' => null,
+                    'context' => [],
+                ];
+            },
+            30
+        );
+
+        $secondPage = $this->query->execute('a_connection_code', $firstPage['search_after'],);
+
+        Assert::assertCount(5, $secondPage['results']);
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQueryIntegration.php:262]: Add comment (30-25 = 5)

**Review by pchasle — APPROVED** [review]

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php [hunk 12]
```diff
@@ -150,90 +150,82 @@ private function buildParameters(?string $encryptedSearchAfter): array
         return $resolver->resolve($parameters);
     }
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/SearchEventSubscriptionDebugLogsQuery.php:152]: Add type array

**Review by pierrejolly — APPROVED** [review]

**Review by tseho — APPROVED** [review]

**Pull request merged**